### PR TITLE
Option to skip exception on cache error

### PR DESCRIPTION
**Pull request by tswayne**

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -7,6 +7,7 @@
 - `name` - determines the name of the cookie used to store session information. Defaults to _session_.
 - `maxCookieSize` - maximum cookie size before using server-side storage. Defaults to 1K. Set to zero to always use server-side storage.
 - `storeBlank` - determines whether to store empty session before they've been modified. Defaults to _true_.
+- `errorOnCacheDisconnect` - will cause yar to throw exception if trying to persist to cache when cache is unavailable. Setting to false will allow application using yar to run uninterrupted if cache is not ready (however sessions will not be saving). Defaults to _true_.
```
Comment (mark-bradshaw) [inline API.md]: I think I'd prefer this be called errorOnCacheNotReady.  I think that's a bit clearer.

Comment (tswayne) [inline API.md]: I agree. Done.

File: lib/index.js [hunk 2]
```diff
@@ -56,8 +57,9 @@ exports.register = function (server, options, next) {
         var load = function () {
 
             request.session = Hoek.clone(request.state[settings.name]);
-            if (request.session &&
-                request.session.id) {
+            var requestSessionSet = request.session && request.session.id;
+            var loadFromCache = settings.errorOnCacheNotReady || cache.isReady();
+            if ( requestSessionSet && loadFromCache) {
```
Comment (mark-bradshaw) [inline lib/index.js]: extra space here

File: lib/index.js [hunk 3]
```diff
@@ -190,8 +192,11 @@ exports.register = function (server, options, next) {
 
     server.ext('onPreResponse', function (request, reply) {
 
+        var skipCacheStorage = !settings.errorOnCacheNotReady && !cache.isReady();
+
         if (!request.session ||
-            (!request.session._isModified && !request.session._isLazy)) {
+            skipCacheStorage ||
```
Comment (mark-bradshaw) [inline lib/index.js]: I'm thinking this may not be placed correctly.  Seems to me we still want to go through the prepare() and cookie() functions.  We just want to skip the storage() function.  Thoughts?

File: test/index.js [hunk 4]
```diff
@@ -547,16 +547,20 @@ it('fails setting session key/value because of failed cache set', { parallel: fa
...
 
+
     var handler = function (request, reply) {
 
         request.session.set('some', 'value');
-        server.stop(function () {                       // Cause cache.set() to fail
-
-            return reply();
-        });
+        return reply();
```
Comment (mark-bradshaw) [inline test/index.js]: Thanks for catching this!

File: test/index.js [hunk 5]
```diff
@@ -575,6 +579,159 @@ it('fails setting session key/value because of failed cache set', { parallel: fa
...
+        expect(err).to.not.exist();
+        server.start(function () {
+
+            server.inject({ method: 'GET', url: '/' }, function (res) {
+
+                var header = res.headers['set-cookie'];
+                var cookie = header[0].match(/(session=[^\x00-\x20\"\,\;\\\x7F]*)/);
+
+                expect(res.statusCode).to.equal(200);
+                expect(handlerSpy.calledOnce).to.equal(true);
```
Comment (mark-bradshaw) [inline test/index.js]: As a preference I'd prefer to avoid adding sinon to the mix.  For better or worse it injects a new methodology of testing than what has been used here.  It's not wrong, but I'd like to keep things consistent.

So, for instance, are you really trying to determine whether the handler got called once, or at all?  If "at all", how about just return a value on line 614 and test for it in the return result?  That's basically what the older tests have been doing.

File: test/index.js [hunk 6]
```diff
@@ -575,6 +579,159 @@ it('fails setting session key/value because of failed cache set', { parallel: fa
...
+
+                cache.prototype.set.restore();
```
Comment (mark-bradshaw) [inline test/index.js]: And for this, can we just add a public method/object on the failing cache mock object that allows us to control how `set` returns?  Again, just to avoid adding another dependency and style to the project.

**Comment by mark-bradshaw** [comment]
That's all I have for now.  I'm in favor of the change.  I think it's valuable.  If we can just pivot a few items then we'll get it in.

File: lib/index.js (hunk 3 above)
Comment (tswayne) [inline lib/index.js]: Good point.  To that point, the check on preResponse is a tad too early too.  I will move that check to the line that checks if the cookie returned session data and add a test for that.

Comment (tswayne) [inline lib/index.js]: Done, the implementation is a bit wordy, so I'll refactor tomorrow when I pull out sinon.

**Comment by mark-bradshaw** [comment]
Thanks.  It's looking good.

File: API.md [hunk 7]
```diff
@@ -7,6 +7,7 @@
 - `name` - determines the name of the cookie used to store session information. Defaults to _session_.
 - `maxCookieSize` - maximum cookie size before using server-side storage. Defaults to 1K. Set to zero to always use server-side storage.
 - `storeBlank` - determines whether to store empty session before they've been modified. Defaults to _true_.
+- `errorOnCacheNotReady` - will cause yar to throw exception if trying to persist to cache when cache is unavailable. Setting to false will allow application using yar to run uninterrupted if cache is not ready (however sessions will not be saving). Defaults to _true_.
```
Comment (hueniverse) [inline API.md]: Change to "Defaults to `false`"

File: lib/index.js [hunk 8]
```diff
@@ -56,10 +57,13 @@ exports.register = function (server, options, next) {
         var load = function () {
 
             request.session = Hoek.clone(request.state[settings.name]);
-            if (request.session &&
-                request.session.id) {
-
+            if ( request.session && request.session.id ) {
```
Comment (hueniverse) [inline lib/index.js]: Don't change style and do it wrong...

File: lib/index.js [hunk 9]
```diff
@@ -191,7 +195,7 @@ exports.register = function (server, options, next) {
     server.ext('onPreResponse', function (request, reply) {
 
         if (!request.session ||
-            (!request.session._isModified && !request.session._isLazy)) {
+           (!request.session._isModified && !request.session._isLazy)) {
```
Comment (hueniverse) [inline lib/index.js]: Again, why change style?

File: test/failing-cache.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
+var internals = {};
```
Comment (hueniverse) [inline test/failing-cache.js]: Test file should match the code file, not add new test files per use case.

File: lib/index.js [hunk 11]
```diff
@@ -248,15 +252,20 @@ exports.register = function (server, options, next) {
 
         var storage = function () {
 
-            reply.state(settings.name, { id: request.session.id });
-            cache.set(request.session.id, request.session._store, 0, function (err) {
+            if (settings.errorOnCacheNotReady || cache.isReady()) {
```
Comment (hueniverse) [inline lib/index.js]: More readable to handle the else first and return.

File: test/index.js (hunk 5 above)
Comment (tswayne) [inline test/index.js]: I have removed sinon and just stubbed methods where needed.

File: API.md (hunk 7 above)
Comment (tswayne) [inline API.md]: Ok, I flipped the default and adjusted the tests.

File: lib/index.js (hunk 8 above)
Comment (tswayne) [inline lib/index.js]: Put back.

File: lib/index.js (hunk 9 above)
Comment (tswayne) [inline lib/index.js]: Put back.

File: lib/index.js (hunk 11 above)
Comment (tswayne) [inline lib/index.js]: Done

File: lib/index.js [hunk 12]
```diff
@@ -60,6 +61,11 @@ exports.register = function (server, options, next) {
                 request.session.id) {
 
                 request.session._isModified = false;
+                if (!settings.errorOnCacheNotReady && !cache.isReady()) {
+                    if (!request.session._store) {
```
Comment (hueniverse) [inline lib/index.js]: The two if statement should be combined into one.

File: API.md (hunk 7 above)
Comment (mark-bradshaw) [inline API.md]: Why flip this to false?  The current behavior matches setting this to true.  I don't want to break expectations.

**Comment by mark-bradshaw** [comment]
Hey Tyler, sorry for the back and forth and some confusion on the defaults.  I would like you to switch the default on errorOnCacheNotReady back to `true`, and then we'll be ready to merge.

File: API.md [hunk 13]
```diff
@@ -7,6 +7,7 @@
 - `name` - determines the name of the cookie used to store session information. Defaults to _session_.
 - `maxCookieSize` - maximum cookie size before using server-side storage. Defaults to 1K. Set to zero to always use server-side storage.
 - `storeBlank` - determines whether to store empty session before they've been modified. Defaults to _true_.
+- `errorOnCacheNotReady` - will cause yar to throw exception if trying to persist to cache when cache is unavailable. Setting to false will allow application using yar to run uninterrupted if cache is not ready (however sessions will not be saving). Defaults to _false_.
```
Comment (hueniverse) [inline API.md]: Where is it throwing an exception? If this is a configurable option, it should explicitly throw or return an error rather than throw because the code fails to protect against a condition.

Comment (mark-bradshaw) [inline API.md]: Tyler's original issue described all this.  The current behavior of yar throws when the cache isn't available.  This option allows yar to continue on without the cache, if desired, and I think that's a useful option.  The current behavior of yar is equivalent to `errorOnCacheNotReady = true`, which is why I want that to default to true.  If you choose to set it to false to avoid the thrown errors, then that's your choice.

Comment (hueniverse) [inline API.md]: Where is it throwing?

Comment (tswayne) [inline API.md]: Error is returned from cache set and get, line 72 and 260 and returned to reply where hapi returns 500.  This option skips those steps to avoid certain exception if the cache is not ready.  When that happens however, the session data does not persist, which is why it shouldn't be default functionality.

**Comment by tswayne** [comment]
I have flipped the default back and cleaned up some test descriptions, so this should be ready to go.

**Comment by mark-bradshaw** [comment]
Thanks Tyler.  I'm going to make one more pass through to digest the final set of changes since we've gone back and forth a bit, and if all's good I'll merge.

**Comment by hueniverse** [comment]
This is an incomplete solution. We need to handle the cache errors regardless of what this option is set to and to handle them correctly. This doesn't actually does what it says but prevent errors from happening.

**Comment by mark-bradshaw** [comment]
I disagree.  I think the option is doing exactly what it says.  If it's set to true, and the cache isn't ready, an error is thrown.  That's exactly what it describes.  If you set this to false, you are choosing to ignore the error.  I don't see the problem.  What do you envision as the correct way to handle the errors?

**Comment by hueniverse** [comment]
It should check if the cache is ready. If it is not, and the option is not set to false, raise an actual error that clearly states the issue. What this is doing is causing a cache issue to fail elsewhere.

**Comment by mark-bradshaw** [comment]
So if the option `errorOnCacheNotReady` is set to false, it should error?  That seems backwards, no?  I can get behind raising a more specific error.  That seems reasonable.  But if you don't care about the cache server availability then you should be able to ignore this condition.  Not every site will have a dedicated ops team that can jump on issues at any time of day or night.  Tyler's patch is a valuable option, so I'll allow it.

@tswayne if you wouldn't mind one final change, if the yar is going to ignore the error, can you please at least do a `request.log` on it so that it can be caught and acted on.  That should serve all our purposes.  You can keep your servers up and running, but we can make sure the error is getting reported so it can be acted on.

**Comment by tswayne** [comment]
@hueniverse if hapi allowed us to catch that specific error and continue on through the request lifecycle (instead of returning 500 response), I would definitely agree that that would be the way to implement this.  Since we have no way to catch the error (no matter how specific it is), handle it (log and NOT return error), and move on, we need to skip cache persistence entirely so we can avoid 500 responses on every request.  This implementation allows users to let yar know that they do not want sessions to crash the application if cache is not available, and I agree with @mark-bradshaw that it is a valuable request.  Additionally "What this is doing is causing a cache issue to fail elsewhere" is why this flag defaults to true by default.  If the user explicitly sets that flag to false they are acknowledging that their sessions will not store or load from cache while their cache is not ready.

@mark-bradshaw I have added request.log

**Comment by hueniverse** [comment]
If the new option is set to false, it will not error. My point is that this solution is a hack, not a proper error handling solution. The problem isn't the new behavior its adding but the existing one it is trying to bypass. We need to first handle the existing error of cache unavailable and surface that better, then we can add the option to ignore it. I am not arguing against the current logic, only that it's an incomplete implementation, now that this issue was raised (which is really a bug).

**Comment by mark-bradshaw** [comment]
Noted.
